File: src/log_migration/authoring/service.py
import json
import shutil
import uuid
from dataclasses import dataclass, replace
from datetime import datetime
from pathlib import Path
from typing import Callable, Literal
from zoneinfo import ZoneInfo

from .database import AuthoringDatabase
from .markdown import RenderedBody, render_markdown
from .models import ConflictError, PageDocument, PublishError, PublishRequest, Redirect, SaveRequest
from .publisher import StaticPublisher
from .repository import ContentRepository, FileTransaction, RepositorySnapshot


TOKYO = ZoneInfo("Asia/Tokyo")
_REDIRECTS_FILENAME = ".authoring-redirects.json"


@dataclass(frozen=True)
class PublishResult:
    page: PageDocument


class AuthoringService:
    def __init__(
        self,
        repository: ContentRepository,
        database: AuthoringDatabase | StaticPublisher,
        publisher: StaticPublisher | Callable[[], datetime],
        clock: Callable[[], datetime] | None = None,
    ):
        if clock is None:
            clock = publisher
            publisher = database
            database = repository.database
        self.repository = repository
        self.database = database
        self.publisher = publisher
        self.clock = clock
        self._redirects: tuple[Redirect, ...] = ()

    def save_draft(self, request: SaveRequest) -> PageDocument:
        page = self.repository.save_draft(request)
        self._refresh()
        return self.repository.get_page(page.id) or page

    def preview(
        self, request: SaveRequest, mode: Literal["local", "public"] = "local"
    ) -> RenderedBody:
        return render_markdown(request.body, mode=mode)

    def publish(self, request: PublishRequest) -> PublishResult:
        snapshot = self._refresh()
        page = self._page(snapshot, request.page_id)
        self._validate_expected_update(page, request)
        now = self._now()
        published = replace(
            page,
            status="published",
            updated_at=now,
            published_at=page.published_at or now,
        )
        candidate = self._replace_page(snapshot, published)
        self._publish_candidate(candidate, page, published)
        return PublishResult(self.repository.get_page(page.id) or published)

    def unpublish(self, page_id: str) -> PageDocument:
        snapshot = self._refresh()
        page = self._page(snapshot, page_id)
        if page.status != "published":
            raise ConflictError("only published pages can be unpublished")
        unpublished = replace(page, status="draft", updated_at=self._now())
        candidate = self._replace_page(snapshot, unpublished)
        self._publish_candidate(candidate, page, unpublished)
        return self.repository.get_page(page.id) or unpublished

    def rename(self, page_id: str, new_name: str) -> PageDocument:
        snapshot = self._refresh()
        if snapshot.problems:
            raise ConflictError("cannot rename while source documents have problems")
        page = self._page(snapshot, page_id)
        source_before = self._source_files() if page.status == "published" else None
        renamed = self.repository.rename_named_page(page_id, new_name)
        if page.status != "published" or renamed.route == page.route:
            self._refresh()
            return renamed

        redirect = Redirect(old_route=page.route, new_route=renamed.route)
        try:
            candidate = self._snapshot_with_redirect(self._refresh(), redirect)
            self._publish_snapshot(candidate, candidate.redirects)
        except Exception:
            assert source_before is not None
            self._restore_source_files(source_before)
            self._refresh()
            raise
        self._refresh()
        return self.repository.get_page(page_id) or renamed

    def _publish_candidate(
        self,
        candidate: RepositorySnapshot,
        original: PageDocument,
        replacement: PageDocument,
    ) -> None:
        staging = self._staging_path()
        try:
            self.publisher.build(candidate, staging)
            transaction = FileTransaction()
            transaction.write(replacement.path, self._serialized(replacement))
            transaction.commit()
            try:
                self.publisher.swap(staging)
            except Exception:
                rollback = FileTransaction()
                rollback.write(original.path, self._serialized(original))
                rollback.commit()
                raise
        except PublishError:
            self._remove_staging(staging)
            self._refresh()
            raise
        except OSError as error:
            self._remove_staging(staging)
            self._refresh()
            raise PublishError(f"could not publish page: {error}") from error
        self._refresh()

    def _publish_snapshot(
        self, snapshot: RepositorySnapshot, redirects: tuple[Redirect, ...]
    ) -> None:
        staging = self._staging_path()
        try:
            self.publisher.build(snapshot, staging)
            transaction = FileTransaction()
            transaction.write(self._redirects_path, self._serialize_redirects(redirects))
            transaction.commit()
            self.publisher.swap(staging)
        except PublishError:
            self._remove_staging(staging)
            raise
        except OSError as error:
            self._remove_staging(staging)
            raise PublishError(f"could not publish page: {error}") from error

    def _refresh(self) -> RepositorySnapshot:
        snapshot = self.repository.refresh()
        self._redirects = self._load_redirects()
        self.database.rebuild(snapshot)
        return replace(snapshot, redirects=self._redirects)

    def _replace_page(
        self, snapshot: RepositorySnapshot, replacement: PageDocument
    ) -> RepositorySnapshot:
        return replace(
            snapshot,
            pages=tuple(
                replacement if page.id == replacement.id else page for page in snapshot.pages
            ),
            redirects=self._redirects,
        )

    def _snapshot_with_redirect(
        self, snapshot: RepositorySnapshot, redirect: Redirect
    ) -> RepositorySnapshot:
        redirects = self._redirects
        if redirect not in redirects:
            redirects = (*redirects, redirect)
        return replace(snapshot, redirects=redirects)

    @staticmethod
    def _page(snapshot: RepositorySnapshot, page_id: str) -> PageDocument:
        page = next((page for page in snapshot.pages if page.id == page_id), None)
        if page is None:
            raise ConflictError(f"page does not exist: {page_id}")
        return page

    @staticmethod
    def _validate_expected_update(page: PageDocument, request: PublishRequest) -> None:
        if request.expected_updated_at is not None and request.expected_updated_at != page.updated_at:
            raise ConflictError("page was updated by another edit")

    def _now(self) -> datetime:
        value = self.clock()
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("clock must return an aware datetime")
        return value.astimezone(TOKYO)

    @staticmethod
    def _serialized(page: PageDocument) -> bytes:
        from .frontmatter import serialize_document

        return serialize_document(page).encode()

    def _staging_path(self) -> Path:
        return self.publisher.output_dir.with_name(
            f"{self.publisher.output_dir.name}.staging-{uuid.uuid4().hex}"
        )

    @staticmethod
    def _remove_staging(path: Path) -> None:
        if path.exists():
            if path.is_dir():
                shutil.rmtree(path)
            else:
                path.unlink()

    def _source_files(self) -> dict[Path, bytes]:
        files = {
            path: path.read_bytes()
            for path in self.repository.content_dir.glob("*.md")
        }
        if self._redirects_path.exists():
            files[self._redirects_path] = self._redirects_path.read_bytes()
        return files

    def _restore_source_files(self, source_before: dict[Path, bytes]) -> None:
        transaction = FileTransaction()
        current_paths = list(self.repository.content_dir.glob("*.md"))
        if self._redirects_path.exists():
            current_paths.append(self._redirects_path)
        for path in current_paths:
            if path not in source_before:
                transaction.delete(path)
        for path, content in source_before.items():
            transaction.write(path, content)
        transaction.commit()

    @property
    def _redirects_path(self) -> Path:
        return self.repository.content_dir / _REDIRECTS_FILENAME

    def _load_redirects(self) -> tuple[Redirect, ...]:
        if not self._redirects_path.exists():
            return ()
        try:
            value = json.loads(self._redirects_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as error:
            raise ConflictError(f"redirect metadata is invalid: {error}") from error
        if not isinstance(value, dict) or set(value) != {"redirects"}:
            raise ConflictError("redirect metadata must contain only redirects")
        redirects = value["redirects"]
        if not isinstance(redirects, list):
            raise ConflictError("redirect metadata redirects must be a list")
        loaded: list[Redirect] = []
        for item in redirects:
            if (
                not isinstance(item, dict)
                or set(item) != {"old_route", "new_route"}
                or not isinstance(item["old_route"], str)
                or not isinstance(item["new_route"], str)
            ):
                raise ConflictError("redirect metadata contains an invalid redirect")
            redirect = Redirect(**item)
            if redirect not in loaded:
                loaded.append(redirect)
        return tuple(loaded)

    @staticmethod
    def _serialize_redirects(redirects: tuple[Redirect, ...]) -> bytes:
        value = {
            "redirects": [
                {"old_route": redirect.old_route, "new_route": redirect.new_route}
                for redirect in redirects
            ]
        }
        return (json.dumps(value, ensure_ascii=False, indent=2, sort_keys=True) + "\n").encode()

File: lib/weblog_authoring/service.rb
# frozen_string_literal: true

require "fileutils"
require "pathname"
require "securerandom"

require_relative "frontmatter"
require_relative "markdown"
require_relative "models"
require_relative "publisher"
require_relative "repository"

module WeblogAuthoring
  class AuthoringService
    attr_reader :repository, :publisher
    attr_writer :publisher

    def initialize(repository, publisher, clock:)
      @repository = repository
      @publisher = publisher
      @clock = clock
      @release_manifest = ReleaseManifest.new(repository.content_dir.join(".authoring-release.json"))
    end

    def save_draft(request)
      repository.save_draft(request)
    end

    def save_and_publish(request)
      saved = save_draft(request)
      publish(PublishRequest.new(page_id: saved.id, expected_updated_at: saved.updated_at))
    end

    def preview(request, mode: "local")
      MarkdownRenderer.new.render(request.body, mode:)
    end

    def today
      current_time.to_date
    end

    def daily_template
      date = today
      weekdays = %w[日曜日 月曜日 火曜日 水曜日 木曜日 金曜日 土曜日]
      {
        title: date.iso8601,
        body: "[[#{weekdays.fetch(date.wday)}]]\n[[#{date.strftime('%Y%m')}]] [[#{date.strftime('%m%d')}]] [[日記]]"
      }
    end

    def publish(request)
      snapshot = repository.refresh
      page = find_page(snapshot, request.page_id)
      validate_expected_update(page, request)
      release_snapshot = load_release_snapshot(snapshot)
      validate_release_sources(snapshot, release_snapshot)
      now = current_time
      published = replace_document(
        page,
        status: "published",
        updated_at: now,
        published_at: page.published_at || now
      )
      candidate = candidate_snapshot(snapshot, published, release_snapshot)
      release_for_build = remove_release_page(release_snapshot, page.id)
      publish_candidate(candidate, release_for_build, published)
      repository.find_route(page.route) || published
    end

    def unpublish(page_id)
      snapshot = repository.refresh
      page = find_page(snapshot, page_id)
      unless page.status == "published"
        raise ConflictError, "only published pages can be unpublished"
      end

      release_snapshot = load_release_snapshot(snapshot)
      validate_release_sources(snapshot, release_snapshot)
      draft = replace_document(page, status: "draft", updated_at: current_time)
      candidate = candidate_snapshot(snapshot, draft, release_snapshot)
      publish_candidate(candidate, release_snapshot, draft)
      repository.find_route(page.route) || draft
    end

    def rename(page_id, new_name)
      snapshot = repository.refresh
      unless snapshot.problems.empty?
        raise ConflictError, "cannot rename while source documents have problems"
      end

      page = find_page(snapshot, page_id)
      release_snapshot = if page.status == "published"
                           loaded = load_release_snapshot(snapshot)
                           validate_release_sources(snapshot, loaded)
                           loaded
                         end
      source_before = source_files if page.status == "published"
      renamed = repository.rename_named_page(page_id, new_name)
      return renamed if page.status != "published" || renamed.route == page.route

      after_rename = repository.refresh
      renamed_release = rename_release_snapshot(release_snapshot, page, renamed)
      redirect = Redirect.new(old_route: page.route, new_route: renamed.route)
      release_after_rename = ReleaseSnapshot.new(
        pages: renamed_release.pages,
        redirects: StaticPublisher.flatten_redirects(renamed_release.redirects + [redirect]),
        published_at: renamed_release.published_at
      )
      candidate = release_candidate_snapshot(after_rename, release_after_rename)
      publish_candidate(candidate, release_after_rename, nil, source_before:)
      repository.find_route(renamed.route) || renamed
    rescue ConflictError, PublishError, ArgumentError, FileTransactionError, SystemCallError, IOError
      restore_source_files(source_before) if defined?(source_before) && source_before
      repository.refresh if defined?(source_before) && source_before
      raise
    end

    private

    def find_page(snapshot, page_id)
      page = snapshot.pages.find { |candidate| candidate.id == page_id }
      raise ConflictError, "page does not exist: #{page_id}" if page.nil?

      page
    end

    def validate_expected_update(page, request)
      return if request.expected_updated_at.nil? || request.expected_updated_at == page.updated_at

      raise ConflictError, "page was updated by another edit"
    end

    def current_time
      value = @clock.call
      raise ArgumentError, "clock must return a Time" unless value.is_a?(Time)

      value.getlocal(TOKYO_OFFSET)
    end

    def load_release_snapshot(snapshot)
      return @release_manifest.load if @release_manifest.path.exist?
      return ReleaseSnapshot.new unless snapshot.pages.any? { |page| page.status == "published" }

      raise PublishError, "release manifest is missing: #{@release_manifest.path}"
    end

    def candidate_snapshot(snapshot, replacement, release_snapshot)
      released_by_id = release_snapshot.pages.each_with_object({}) do |page, pages|
        pages[page.id] = page
      end
      pages = snapshot.pages.map do |page|
        next replacement if page.id == replacement.id

        released_by_id.fetch(page.id, page)
      end
      RepositorySnapshot.new(
        pages:,
        problems: snapshot.problems,
        redirects: release_snapshot.redirects
      )
    end

    def release_candidate_snapshot(snapshot, release_snapshot)
      released_by_id = release_snapshot.pages.each_with_object({}) do |page, pages|
        pages[page.id] = page
      end
      RepositorySnapshot.new(
        pages: snapshot.pages.map { |page| released_by_id.fetch(page.id, page) },
        problems: snapshot.problems,
        redirects: release_snapshot.redirects
      )
    end

    def validate_release_sources(snapshot, release_snapshot)
      current_by_id = snapshot.pages.each_with_object({}) do |page, pages|
        pages[page.id] = page
      end
      problems_by_path = snapshot.problems.each_with_object({}) do |problem, problems|
        problems[problem.path] = problem
      end

      release_snapshot.pages.each do |released_page|
        problem = problems_by_path[released_page.path]
      unless problem.nil?
        raise PublishError, "released page is invalid: #{released_page.route}: #{problem.detail}"
      end

        current_page = current_by_id[released_page.id]
        raise PublishError, "released page is missing: #{released_page.route}" if current_page.nil?
        next if current_page.path == released_page.path

        raise PublishError, "released page path changed: #{released_page.route}"
      end
    end

    def remove_release_page(snapshot, page_id)
      ReleaseSnapshot.new(
        pages: snapshot.pages.reject { |page| page.id == page_id },
        redirects: snapshot.redirects,
        published_at: snapshot.published_at
      )
    end

    def rename_release_snapshot(snapshot, original, renamed)
      pages = snapshot.pages.map do |page|
        body = WeblogAuthoring.replace_wiki_links(
          page.body,
          old_name: original.name.to_s,
          new_name: renamed.name.to_s
        )
        changes = { body:, links: WeblogAuthoring.extract_wiki_links(body) }
        if page.id == original.id
          changes[:name] = renamed.name
          changes[:path] = renamed.path
        end
        PageDocument.new(**page.to_h.merge(changes))
      end
      ReleaseSnapshot.new(
        pages:,
        redirects: snapshot.redirects,
        published_at: snapshot.published_at
      )
    end

    def publish_candidate(snapshot, release_snapshot, replacement, source_before: nil)
      staging = repository.content_dir.dirname.join(
        "#{publisher.output_dir.basename}.staging-#{SecureRandom.hex(8)}"
      )
      source_before ||= source_files

      begin
        result = publisher.build(snapshot, staging, release_snapshot:)
        transaction = FileTransaction.new
        unless replacement.nil?
          transaction.write(replacement.path, WeblogAuthoring.serialize_document(replacement))
        end
        transaction.write(@release_manifest.path, result.release_manifest_json)
        transaction.commit

        begin
          publisher.swap(staging)
        rescue PublishError, FileTransactionError, SystemCallError, IOError
          restore_source_files(source_before)
          raise
        end
      rescue PublishError
        restore_source_files(source_before)
        remove_staging(staging)
        repository.refresh
        raise
      rescue ArgumentError, TypeError, FileTransactionError, SystemCallError, IOError => error
        restore_source_files(source_before)
        remove_staging(staging)
        repository.refresh
        raise PublishError, "could not publish page: #{error.message}"
      ensure
        remove_staging(staging)
      end

      repository.refresh
    end

    def source_files
      files = Pathname.glob(repository.content_dir.join("*.md").to_s).each_with_object({}) do |path, result|
        result[path] = path.binread
      end
      files[@release_manifest.path] = @release_manifest.path.binread if @release_manifest.path.exist?
      files
    end

    def restore_source_files(files)
      transaction = FileTransaction.new
      current_paths = Pathname.glob(repository.content_dir.join("*.md").to_s)
      current_paths << @release_manifest.path if @release_manifest.path.exist?
      current_paths.uniq.each do |path|
        transaction.delete(path) unless files.key?(path)
      end
      files.each { |path, content| transaction.write(path, content) }
      transaction.commit
    end

    def remove_staging(path)
      FileUtils.rm_rf(path) if path.exist?
    end

    def replace_document(document, **changes)
      PageDocument.new(**document.to_h.merge(changes))
    end
  end
end
